flow/project.py

```diff
@@ -1601,13 +1870,31 @@ def _generate_operations(self, cmd, jobs, requires=None):
             cmd_ = cmd.format(job=job)
             yield JobOperation(name=cmd_.replace(' ', '-'), cmd=cmd_, job=job)
 
-    def _get_pending_operations(self, jobs, operation_names=None):
+    def _get_pending_groups(self, jobs, groups, exec_mode=False):
+        if exec_mode:
+            mode = 'exec'
+        else:
+            mode = 'run'
+        for job in jobs:
+            for group in groups:
+                if group.eligible(job):
+                    yield group.create_job_operation(job, mode=mode, index=0)
+
+    def _get_pending_operations(self, jobs, operation_names=None,
+                                exec_mode=False):
         "Get all pending operations for the given selection."
         assert not isinstance(operation_names, six.string_types)
-        for op in self.next_operations(* jobs):
+        for op in self.next_operations(* jobs, exec_mode=exec_mode):
             if operation_names is None or any(fullmatch(n, op.name) for n in operation_names):
                 yield op
 
+    def _verify_group_compatibility(self, groups):
```

Comment:
I'd suggest to inline this and express it more compactly:
`not any(a.intersection(b) for a, b in zip(groups[1:], groups[:-1]))`

Reply:
This would not correctly iterate over pairs. I suggest
`all(a.compatible(b) for a, b in itertools.combinations(groups, 2))` or
`not any(a.intersection(b) for a, b in itertools.combinations(groups, 2))`.
Because A intersection B and B intersection C both being the null set does not imply that A intersection C is the null set.

I also think the function adds some readability though in my first suggestion not much.

Reply:
Oh, you are correct, sorry about that. I'd still like to switch to using built-in set functions (`intersection()`) instead of the `compatible()`class method.

Here are my arguments:

1. Making the `FlowGroup` class iterable over its members provides value beyond that one check.
2. It is not well-defined what `compatible()` means. Just because two groups are incompatible in this very specific context, does not mean that they are incompatible under all circumstances.

If you are concerned about obfuscating the code, there are other (IMO better) ways to express this, for example with a variable name (`compatible = ...`) or by keeping the dedicated function here (`_verify_group_compatibility`), just not on the `FlowGroup` class level.

Reply:
Currently the only context in which compatibility is ambigious is in the case where one is using groups directly in the run command. Otherwise, compatibility is consistent in meaning. Also the class is already iterable, but just keeping the logic in `_verify_group_compatibility` I think is a good solution as one would be able to quickly see what contexts compatibility is defined for.